### PR TITLE
Set data for child forms and create named forms.

### PR DESCRIPTION
**Pull request by tormit**
Not sure if this was bug or missing feature, but I was not able to pass data for child forms. Now its possible.
Also added new method in FormBuilder to create named forms like Symfony does. Form field names become like youname[fieldname].

### CODE REVIEW

**Comment by kristijanhusak** [comment]
How were you trying to pass data to child forms? Can you give me some code examples?

Creating named forms is great idea, but it can be done in another way, much simpler.
If you check [here](https://github.com/kristijanhusak/laravel-form-builder/blob/master/src/Kris/LaravelFormBuilder/Fields/ChildFormType.php#L72), you will see that you just need to pass `is_child` flag, and name that form needs to carry. So basically this is enough:
```php
    $form = FormBuilder::plain([
        'is_child' => true,
        'name' => 'yourname',
        'model' => $model
    ])->add('name')->add('email', 'email');
```

and you get `yourname[name]` and `yourname[email]`

The only problem here is that if you pass in model, it's data needs to be structured in that way:
```php
$model = [
    'somefield' => 'sometext',
    'yourname' => [ 'name' => 'example name', 'email' => 'test@example.com']
];
```

In the background model binding is done by Laravel's form builder class, so this is something that i'm not sure can be fixed, maybe just to check if not passed like that, and then to wrap it.

**Comment by tormit** [comment]
The problem with 
```php
   $form = FormBuilder::plain()
```
is that it creates only plain form, but I need to create named form from pre-defined form.
And the issue about passing to child form was related to this named forms case.

The idea of createNamedForm() is to be like it is in Symfony. 
```php
   // form builder in Symfony
   $form = $this->createForm(
                $this->get('my.form.service'),
                $entity,
                [
                    'attr' => []
                ]
            )
```
Symfony FormBuilder creates named forms out of box.

Another possibility to make your FormBuilder similar to Symfony is to enable user to implement _getName()_ method in form definition class. If user has implemented _getName()_, then _FormBuilder::create()_ automatically created named form.

**Comment by kristijanhusak** [comment]
You can do same thing with existing form:
```php
namespace App\Forms;

class MyForm extends Form {
    public function buildForm() 
    {
        $this->add('name')->add('email', 'email');
    }
}

// Controller
 $form = FormBuilder::create('App\Forms\MyForm', [
        'is_child' => true,
        'name' => 'yourname',
        'model' => $model
    ]);
```

Wrapper for this definitely needs to be created.

About `getName`, i would not implement that because there is already a property that can be set (`$childFormName`), but it will not work at this moment, so I need to fix that, and i'll probably skip `is_child` flag also. That way we will be able just to put property for example `$name` on a form, and any time it is created, it will add that.

**Comment by kristijanhusak** [comment]
I just updated it to use only the `name` key. More info can be found [here](https://github.com/kristijanhusak/laravel-form-builder#named-forms). Form builder method for this is still needed, so fix your PR and i'll merge it.

**Comment by tormit** [comment]
Cool. I'll do it tomorrow.

**Comment by tormit** [comment]
Using you update, I'm able to create named forms, but passing existing data still won't work. My patch does not work anymore either. I'll try to solve it.

**Comment by kristijanhusak** [comment]
Yeah, as i said, it expects data to be in that namespace, so if you pass `['name' => 'john', 'email' => 'johndoe@example.com']`, it needs to get it in this form:
```php
$data = [
    'form_name' => [
        'name' => 'johndoe',
        'email' => 'johndoe@example.com'
    ]
];
```

**Comment by tormit** [comment]
I updated my comment with my code.

**Comment by kristijanhusak** [comment]
3rd parameter for `create` method is used for passing data to the form class, but that data will not be used in the form itself. For that model is needed. Here's short explanation of using data:

```php
$data = ['some' => 'data passed for usage'];

$model = ['name' => 'johndoe', 'email' => 'johndoe@example.com'];

   $form = $formBuilder->create('\App\Forms\Customer\CompanyProfileForm', [
            'name' => 'profile',
            'method' => 'POST',
            'url' => route('customer.company.profile'),
            'model' => $model
        ], $data);


class CompanyProfileForm extends Form {
    public function buildForm()
    {
        $data = $this->getData('some'); // 'data passed for usage'
        $model = $this->getModel(); // Passed model, will be bound to form with laravels Form::model()
        $this->add('name')->add('email');
    }
}
```

This way it should work:
```php
$model = ['name' => 'johndoe', 'email' => 'johndoe@example.com'];

   $form = $formBuilder->create('\App\Forms\Customer\CompanyProfileForm', [
            'name' => 'profile',
            'method' => 'POST',
            'url' => route('customer.company.profile'),
            'model' => ['profile' => $model]
        ], $data);
```
Only problem is if the data is already namespaced in that way, there will be a problem. So some kind of check needs to be done.

**Comment by tormit** [comment]
OK. Now I understand difference between data and model. 
The model object is present in CompanyProfileForm::buildForm(), but still no data is bound to form.

**Comment by kristijanhusak** [comment]
Can you please give me the dump of data from the model that is passed?

**Comment by tormit** [comment]
http://postimg.org/image/nt6x38n5v/e8ac4762/

**Comment by kristijanhusak** [comment]
Thanks. As i see, you do not have any attributes with 'profile' key. So try doing this:
```php
$form = $formBuilder->create('\App\Forms\Customer\CompanyProfileForm', [
            'name' => 'profile',
            'method' => 'POST',
            'url' => route('customer.company.profile'),
            'model' => ['profile' => $model->toArray()]
        ], $data);
```

**Comment by tormit** [comment]
Now it works. But I think it should be simpler. A developer would assume that given model automatically applies to the named form.

**Comment by kristijanhusak** [comment]
Yeah, i know. That needs to be handled through the FormBuilder named form method that you want to add. Basically it should be enough to check if the model is passed, and if it's eloquent, to convert it to array, and then namespace it by the provided name, like you done it when passing the model.

**Comment by tormit** [comment]
Check my update. I removed createNamedForm() as it is irrelevant now. Also added way to "fix" model for named forms.

File: src/Kris/LaravelFormBuilder/FormBuilder.php [hunk 1]
```diff
@@ -48,6 +50,15 @@ public function create($formClass, array $options = [], array $data = [])
             ->setFormOptions($options)
             ->addData($data);
 
+        // reset model to match named form logic
+        if ($form->getName()) {
```
Comment (kristijanhusak) [inline src/Kris/LaravelFormBuilder/FormBuilder.php]: This logic can be moved maybe [here](https://github.com/kristijanhusak/laravel-form-builder/blob/master/src/Kris/LaravelFormBuilder/Form.php#L543).

File: src/Kris/LaravelFormBuilder/FormBuilder.php [hunk 2]
```diff
@@ -48,6 +50,15 @@ public function create($formClass, array $options = [], array $data = [])
...
+            if ($form->getModel() instanceof Model && method_exists($form->getModel(), 'toArray')) {
+                $form->setModel([$form->getName() => $form->getModel()->toArray()]);
+            } elseif (is_array($form->getModel())) {
```
Comment (kristijanhusak) [inline src/Kris/LaravelFormBuilder/FormBuilder.php]: We also need to check if model maybe already have that namespace, so we do not double it.

Comment (tormit) [inline src/Kris/LaravelFormBuilder/FormBuilder.php]: Ok. I'll add fix soon.

**Comment by tormit** [comment]
I made the fix. But I didn't understand how you want this logi to be moved.
You may take my changes and place this logic somewhere you consider more suitable.

**Comment by kristijanhusak** [comment]
Ok, I will. Thanks for the help and idea for this!

**Comment by tormit** [comment]
I see you have implemented model binding in your code. I assume that I may switch my project back to your version?
I'm closing this pull request.

**Comment by kristijanhusak** [comment]
@tormit 
Yes, i did, but still didn't tag a release, want to add few more small things. Will tag it today probably.

**Comment by tormit** [comment]
Ok. Thanks. This is great plugin. This is a missing feature of Laravel for those, like me, coming from Symfony.
Keep up good work!

**Comment by kristijanhusak** [comment]
It's done, do a composer update :)

**Pull request closed without merging**